### PR TITLE
Basic accounting for threads

### PR DESCRIPTION
**Pull request by czapiga**
These changes implements #246

### CODE REVIEW

**Comment by czapiga** [comment]
After NetBSD code analysis I think there is little change needed here. I will update this ASAP

**Comment by cahirwpz** [comment]
@rafalcieslak Could you investigate what caused a Travis failure – it's somehow related to `run_tests.py` script...

**Comment by rafalcieslak** [comment]
All it says is that kernel output had a non-ascii byte. That means either there are some non-ascii characters in our source code (some text editors automatically perform fancy substitutions for some characters), or kernel started dumping binary data (unterminated string, invalid ptr). I'm not sure if that's easily fixable by improving `run_tests.py`, I don't know any good way to dump raw binary data to std output in Python.

**Comment by rafalcieslak** [comment]
I hope I'll be able to tell more about this problem once #304 is merged.

**Comment by rafalcieslak** [comment]
I've found the reason for non-ascii characters in output log:

```
Running test ramdisk.
file /usr/include/sys/errno.h:
/* errno is not a global variable, because that would make using it
   non-reentrant.  Instead, its address is returned by the function
   __errno.  */

#ifndef _SYS_ERRNO_H_
#ifdef __cplusplus
extern "C" {
#endif
#define _SYS_ERRNO_H_

#include <sys/reent.h>

#ifndef _REENT_ONLY
#define errno (*__errno())
extern int *__errno _PARAMS ((void));
#endif

/* Please don't use these variables directly.
   Use strerror instead. */
extern __IMPORT _CONST char * _CONST _sys_errlist[];
extern __IMPORT intPÂÂ€
```

Apparently the ramdisk tests reads some invalid data from include headers used for demonstration. I'll open a separate issue addressing the problem.

I have no idea what causes the test to fail, though. `ktest_failure()` does get called, but it gets stuck in
```c
    for (test_entry_t **ptr = autorun_tests; *ptr != NULL; ptr++) {
```
??? `autorun_tests` should never be modified after tests have started, so it looks like some major memory corruption (or Travis output is bonkers).

**Comment by cahirwpz** [comment]
Blocked by #331 as we need to measure time with higher resolution for the tests to pass.

**Comment by czapiga** [comment]
I don't understand travis output.. I did run `make format` and I'm aligned with upstream/master..

**Comment by czapiga** [comment]
@rafalcieslak please, can you look at this from travis POV

File: tests/thread_stats.c [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  }
+  for (int i = 0; i < THREADS_NUMBER; i++) {
+    thread_join(threads[i]);
+  }
+  for (int i = 0; i < THREADS_NUMBER; i++) {
+    klog("Thread:%d runtime:%u.%u sleeptime:%u.%u context switches:%llu", i,
+         threads[i]->td_rtime.tv_sec, threads[i]->td_rtime.tv_usec,
+         threads[i]->td_slptime.tv_sec, threads[i]->td_slptime.tv_usec,
+         threads[i]->td_nctxsw);
+    if ((threads[i]->td_rtime.tv_sec == 0 &&
```
Comment (cahirwpz) [inline tests/thread_stats.c]: Could you rewrite it in terms of `timeval_isset`?

File: tests/thread_stats.c [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#include <thread.h>
+#include <ktest.h>
+#include <klog.h>
+#include <sched.h>
+#include <runq.h>
+
+#define THREADS_NUMBER 10
+#define TEST_TIME 500
```
Comment (cahirwpz) [inline tests/thread_stats.c]: 500 µs? 500ms? 500s? Please use `timeval_t` values to explicitely express time unit:
```c
const timeval_t test_time = TIMEVAL(0.5);
```

File: tests/thread_stats.c [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+static void thread_nop_function(void *arg) {
+  while (tv2st(get_uptime()) < TEST_TIME + tv2st(*(timeval_t *)arg))
```
Comment (cahirwpz) [inline tests/thread_stats.c]: Why so complicated? Isn't it just better to calculate everything using `timeval_t` type?

**Comment by cahirwpz** [comment]
@czapiga which version of `clang-format` do you use locally?

**Comment by cahirwpz** [comment]
@czapiga If you can reformulate `timeval_t` API to make it more convenient to use and render code that uses it more concise, then go for it!

**Comment by czapiga** [comment]
@cahirwpz I use clang-format 3.9.1

**Comment by czapiga** [comment]
I think tests can sometimes fail due to bug found during  #342. I will try to finish  #342 ASAP.

**Comment by czapiga** [comment]
@cahirwpz I think it's ready to merge

**Pull request merged**